src/wayland/compositor/mod.rs

```diff
@@ -415,8 +417,9 @@ pub fn add_post_commit_hook(surface: &WlSurface, hook: fn(&DisplayHandle, &WlSur
 /// Register a destruction hook to be invoked on surface destruction
 ///
 /// It'll be invoked when the surface is destroyed (either explicitly by the client or on
-/// client disconnect).
-pub fn add_destruction_hook(surface: &WlSurface, hook: fn(&SurfaceData)) {
+/// client disconnect). The compositor state is passed back as `Any` and may be downcast as needed
+/// by the callee.
+pub fn add_destruction_hook(surface: &WlSurface, hook: fn(&mut dyn Any, &SurfaceData)) {
```

Comment:
Could smithay provide a thunk so under the hood it's &mut dyn Any for preventing further generic expansion inside the implementation but still make the api type safe?

Reply:
Ideally smithay would track a list of closures. This function would add a closure to the internal surface destroy hook list. This internal closure would just downcast Any to D (hint 'static bound) and then call whatever function was provided.

Reply:
Pasting relevant IRC conversation
```rust
let thunk = Box::new(|dyn_data, surface_data| {
    destroy_callback(dyn_data.downcast_mut::<D>().expect("generic on add_destruction_hook does not match dispatched data"), surface_data);
});

// Add the type erased closure to the destructor hooks list as a bunch of Box<Fn(...)>
```